Fix Scene hot reloading.

andriyDev opened the pull request:
```
This is a draft until #18357 is resolved - otherwise, hot-reloading scenes would leave them completely mangled.
```

# Objective

- Fixes #14698

## Solution

- Refactor all the `SceneSpawner` functions to include a `_dynamic` suffix where appropriate.
- Add non-DynamicScene versions of functions on `SceneSpawner`.
- Listen for `Scene` `AssetEvent`s and update the corresponding spawned scenes.

## Testing

- Ran the `hot_asset_reloading` example and mutated the GLTF to see it be updated
- TODO: Make sure this works properly after #18357 is fixed.

---

## Migration Guide

- Some of the methods on `SceneSpawner` have been renamed:
    - `despawn` -> `despawn_dynamic`
    - `despawn_sync` -> `despawn_dynamic_sync`
    - `update_spawned_scenes` -> `update_spawned_dynamic_scenes`
- In their place, we've also added `despawn`, `despawn_sync`, and `update_spawned_scenes` which all act on `Scene`s (as opposed to `DynamicScene`s).

andriyDev commented:
Looks like this PR isn't quite ready. I've got two meshes. Adding an empty node and saving after that causes a panic:

```
thread 'Compute Task Pool (0)' panicked at crates\bevy_transform\src\systems.rs:466:21:
assertion `left == right` failed
  left: 11v1#4294967307
 right: 12v1#4294967308
```

This also causes the entire app to freeze and blast my CPU usage to 50% (IIRC half the threads are for systems and half for assets, so I guess all the system threads are blasting).

Initial investigation shows the hierarchy is not set up correctly, but I'm not sure why yet.

schrottkatze commented:
Hi, uh, are you planning on continuing to work on this?
(if you forgot about this, i guess this comment could serve as a ping like "yes people do still need this" ^^', don't mean to be like passive-aggressive, i've just spent like an hour thinking i've been doing something wrong just to realize it's broken in bevy, actually and the only pr to fix it seems stalled >~<)

andriyDev commented:
Hi, so sorry I never actually went and updated the status here. 

This is somewhat blocked on #18418. I say somewhat because it's not clear to me that entities in a scene should be reused at all (which is what that issue is about).

Entities in a Scene do not have any semantic information. That means when you change the scene, the entities before and after have no correspondence. Imagine you have the following Scene (which comes from loading a GLTF)

```
Entity 1v0: Cube
Entity 2v0: Sphere
```

If you go and delete the cube, you get the Scene: 

```
Entity 1v0: Sphere
```

Note that the sphere now uses the cube's Entity. That means hot reloading would try to turn the cube entity into the sphere. Any components the cube had that the sphere doesn't will **not** be replaced (for example, if the cube had an IAmAnItemPickup component, now the sphere does. What if the sphere was actually the floor of your world!). This whole problem is exacerbated by relations. 

I think one approach is to just delete all the entities and spawn new ones. It makes hot reloading less powerful, but makes it behave more reasonably IMO. I can try drafting that up soon. The alternative is this being blocked on #18418 haha.

andriyDev commented:
Actually, maybe even that alternative is a non-starter since if you had something parented to your scene, it would be despawned when you despawned all your entities... So this PR might just be blocked...

andriyDev commented:
Alright, after some discussion in Discord, I've been convinced that deleting the scene and respawning it is better than doing nothing. I also added a test for scenes and dynamic scenes to make sure they continue to hot-reload.

Note: I've fixed the rendering change, so we don't need the label!

janhohenheim commented on crates/bevy_scene/src/scene_spawner.rs:
```diff
@@ -286,7 +310,29 @@ impl SceneSpawner {
+                    }
+                }
+            }
+        }
```
Nothing blocking, but this could be simplified a lot by using `.filter_map` for the spawned scenes, followed by `.flat_map` for the spawned instances, followed again by `filter_map` for the instance info ;)

andriyDev replied:
I would prefer not to do this to keep it consistent with the dynamic one, and I'd prefer not to change that one to keep the diff small. bevy_scene could use a style pass IMO

janhohenheim reviewed (approved):
Looked through all commits and everything looks correct to me :)
Ping @Noxmore

andriyDev commented:
@ChristopherBiscardi, could I see your observer code? Is it possible you're spawning an entity instead of adding a component? Hotreloading literally despawns every entity associated with the scene, so I don't know what entities your component could be living off of?

bevy_inspector_egui might also shed some light on this. I mean if the entire hierarchy is duplicated after hot reload, then we'll know this PR isn't working.

ChristopherBiscardi commented:
yeah that was it, I confirmed with a Local that prevented the spawn if already had been spawned. reloading itself is working fine.
